pennylane/utils.py

```diff
@@ -163,3 +166,110 @@ def expand(U, wires, num_wires):
 
     # permute U to take into account rearranged wires
     return U[:, perm][perm]
+
+
+class Recorder:
+    """Recorder class used by the :class:`~.OperationRecorder`.
+
+    The Recorder class is a very minimal QNode, that simply
+    acts as a QNode context for operator queueing."""
+    # pylint: disable=too-few-public-methods
+    def __init__(self, old_context):
+        self._old_context = old_context
+        self._ops = []
+
+    def _append_op(self, op):
+        """:class:`~.Operator` objects call this method
+        and append themselves upon initialization."""
+        self._ops.append(op)
+
+        # this ensure the recorder does not interfere with
+        # any QNode contexts
+        if self._old_context:
+            self._old_context._append_op(op)
+
+    @property
+    def queue(self):
+        """Queue of the underlying QNode if existant, otherwise an empty list."""
+        if self._old_context:
+            return self._old_context.queue
+        
+        return []
+
+    # Spoof all attributes of the underlying QNode if there is one
+    def __getattr__(self, name):
+        if self._old_context:
+            return self._old_context.__getattribute__(name)
+
+        raise AttributeError("Attribute {} of Recorder mock QNode does not exist.".format(name))
```

Comment:
The idea behind this was that if there was a QNode in the `_current_context`, one should not realize that the Recorder was put in between. E.g., if you want to call `QNode._current_context.device`. In order not to have to reproduce all properties and methods of the QNode I just wrote this bit to relay every requested but unknown attribute directly to the underlying QNode.

I also have to admit that I wanted to learn how to do this, because I always wondered how `MagicMock` is doing it.

Reply:
There remains the question what happens if only the Recorder is present and there is no underlying QNode. Up to now an error is raised if things are attempted that are not possible.